staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go

```diff
@@ -42,63 +46,130 @@ func (r *versionDiscoveryHandler) ServeHTTP(w http.ResponseWriter, req *http.Req
+		}
+
+		apiResourcesForDiscovery = append(apiResourcesForDiscovery, metav1.APIResource{
+			Name:               crd.Status.AcceptedNames.Plural,
+			SingularName:       crd.Status.AcceptedNames.Singular,
+			Namespaced:         crd.Spec.Scope == apiextensionsv1.NamespaceScoped,
+			Kind:               crd.Status.AcceptedNames.Kind,
+			Verbs:              verbs,
+			ShortNames:         crd.Status.AcceptedNames.ShortNames,
+			Categories:         crd.Status.AcceptedNames.Categories,
+			StorageVersionHash: storageVersionHash,
+		})
+
+		subresources, err := apiextensionshelpers.GetSubresourcesForVersion(crd, requestedVersion)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		if subresources != nil && subresources.Status != nil {
+			apiResourcesForDiscovery = append(apiResourcesForDiscovery, metav1.APIResource{
+				Name:       crd.Status.AcceptedNames.Plural + "/status",
+				Namespaced: crd.Spec.Scope == apiextensionsv1.NamespaceScoped,
+				Kind:       crd.Status.AcceptedNames.Kind,
+				Verbs:      metav1.Verbs([]string{"get", "patch", "update"}),
+			})
+		}
+
+		if subresources != nil && subresources.Scale != nil {
+			apiResourcesForDiscovery = append(apiResourcesForDiscovery, metav1.APIResource{
+				Group:      autoscaling.GroupName,
+				Version:    "v1",
+				Kind:       "Scale",
+				Name:       crd.Status.AcceptedNames.Plural + "/scale",
+				Namespaced: crd.Spec.Scope == apiextensionsv1.NamespaceScoped,
+				Verbs:      metav1.Verbs([]string{"get", "patch", "update"}),
+			})
+		}
 	}
-	ret, ok := clusterDiscovery[gv]
-	return ret, ok
-}
 
-func (r *versionDiscoveryHandler) setDiscovery(clusterName string, gv schema.GroupVersion, discoveryHandler *discovery.APIVersionHandler) {
-	r.discoveryLock.Lock()
-	defer r.discoveryLock.Unlock()
+	resourceListerFunc := discovery.APIResourceListerFunc(func() []metav1.APIResource {
+		return apiResourcesForDiscovery
+	})
 
-	if _, clusterExists := r.discovery[clusterName]; !clusterExists {
-		r.discovery[clusterName] = map[schema.GroupVersion]*discovery.APIVersionHandler{}
-	}
-	r.discovery[clusterName][gv] = discoveryHandler
-}
+	// HACK: if we are adding resources in legacy scheme group through CRDs (KCP scenario)
+	// then do not expose the CRD `APIResource`s in their own CRD-related group`,
+	// But instead add them in the existing legacy schema group
+	// if genericcontrolplanescheme.Scheme.IsGroupRegistered(clusterGroupVersion.Group) {
+	// 	if !foundGroup || !foundVersion {
+	// 		delete(discovery.ContributedResources, clusterGroupVersion)
+	// 	}
 
-func (r *versionDiscoveryHandler) unsetDiscovery(clusterName string, gv schema.GroupVersion) {
-	r.discoveryLock.Lock()
-	defer r.discoveryLock.Unlock()
+	// 	discovery.ContributedResources[clusterGroupVersion] = resourceListerFunc
+	// 	return nil
+	// }
 
-	if _, clusterExists := r.discovery[clusterName]; !clusterExists {
+	if !foundGroup || !foundVersion {
+		r.delegate.ServeHTTP(w, req)
```

Comment:
I think this was the existing practice - if you can't handle something, just keep sending it down the line to the delegate. I'm guessing this would allow for consistent 404 responses? (delegate is `genericapiserver.NewEmptyDelegate()` in this particular case).